library/init/meta/uneval.lean

```diff
@@ -0,0 +1,74 @@
+/-
+Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Ullrich
+-/
+prelude
+import init.meta.tactic
+
+open tactic
+
+meta class has_uneval (α : Type) :=
+(uneval : α → expr)
+
+@[inline] meta def uneval {α : Type} [has_uneval α] : α → expr :=
+has_uneval.uneval
+
+private meta def c (n : name) := expr.const n []
+
+-- TODO: use macro
+private meta def nat.uneval : nat → expr
+| n :=
+  if n = 0 then c ``nat.zero
+  else
+    let f := if n % 2 = 0 then ``bit0 else ``bit1 in
+    c f (nat.uneval (n / 2))
+
+meta instance : has_uneval nat := ⟨nat.uneval⟩
+
+meta instance : has_uneval char :=
+⟨λ ⟨n, pr⟩, expr.app (expr.const ``char.of_nat []) $ uneval n⟩
+
+-- TODO: use macro
+private meta def string.uneval : string → expr
+| []      := c ``string.empty
+| (d::cs) := c ``string.str (uneval d) (string.uneval cs)
+
+meta instance : has_uneval string := ⟨string.uneval⟩
+
+meta instance : has_uneval unsigned :=
+⟨λ ⟨n, pr⟩, c ``unsigned.of_nat (uneval n)⟩
+
+meta def name.uneval : name → expr
+| name.anonymous        := c ``name.anonymous
+| (name.mk_string s n)  := c ``name.mk_string  (uneval s) (name.uneval n)
+| (name.mk_numeral i n) := c ``name.mk_numeral (uneval i) (name.uneval n)
+
+meta instance : has_uneval name := ⟨name.uneval⟩
+
+meta instance : has_uneval expr  := ⟨expr.quote⟩
+meta instance : has_uneval pexpr := ⟨expr.quote ∘ pexpr.to_raw_expr⟩
+
+private meta def list.uneval {α : Type} [has_uneval α] : list α → expr
+| []     := c ``list.nil
+| (h::t) := c ``list.cons (uneval h) (list.uneval t)
+
+meta instance {α : Type} [has_uneval α] : has_uneval (list α) := ⟨list.uneval⟩
+
+meta instance {α : Type} [has_uneval α] : has_uneval (option α) :=
+⟨λ opt, match opt with
+| some x := c ``some (uneval x)
+| none   := c ``none
+end⟩
+
+meta instance : has_uneval unit := ⟨λ _, c ``unit.star⟩
+
+meta instance {α β : Type} [has_uneval α] [has_uneval β] : has_uneval (α × β) :=
+⟨λ ⟨x, y⟩, c ``prod.mk (uneval x) (uneval y)⟩
+
+-- TODO: no worky
+-- meta def nat.to_expr : nat → tactic expr := λ n, to_expr `(%%(uneval n) : nat)
+-- meta def char.to_expr : char → tactic expr := λ n, to_expr `(%%(uneval n) : char)
+-- meta def unsigned.to_expr : unsigned → tactic expr := λ n, to_expr `(%%(uneval n) : unsigned)
+-- meta def name.to_expr : name → tactic expr := λ n, to_expr `(%%(uneval n) : name)
+-- meta def list_name.to_expr : list name → tactic expr := λ n, to_expr `(%%(uneval n) : list name)
```

Comment:
I think `` `(%%(uneval n) : nat) `` treats `uneval n` as an (elaborated) expression, since it has the type `expr`, and obviously won't elaborate it then.  Hence the empty universe parameter list in `list.nil` for example will never be filled.